python/cudf/cudf/utils/utils.py

```diff
@@ -280,6 +281,33 @@ def __get__(self, instance, cls):
             return value
 
 
+class GetAttrGetItemMixin:
+    """This mixin changes `__getattr__` to attempt a `__getitem__` call.
+
+    Subclasses may define a `_PROTECTED_KEYS` set as a class attribute to
+    indicate any keys that may be accessed by __getitem__. Testing for these
+    keys allows this class to prevent recursion errors when subclasses are
+    copied (see below for more details). Note that a cleaner solution for this
+    problem would be to override one of the serialization protocols (e.g.
+    `__reduce__` or `__setstate__`, but this class may be used in complex
+    multiple inheritance hierarchies that might also override serialization.
+    The solution here is a minimally invasive change to simplify such patterns.
+    """
```

Comment:
Great explanation, though I think some of the comments can be moved to the PR description page. For documentation purpose, this may reveal too much implementation (plus non-implementation) detail. The mentioning of "recursion prevention" is duplicate of that below.

I would, however, like to know more about the `__reduce__` and `__setstate__` solution. Could you explain how that might be done?

Reply:
Good points. This class is entirely internal and the docstring is intended to be for developers, so some amount of implementation detail is intentional, but I'm not sure what the right balance is. What do you think?

Reply:
Your `__setstate__` question is a much more complicated one.

Short version: you can change how an object is copied (via `copy.copy`) and serialized (`via pickle.dump`) using `__setstate__`, so if you implement those you can implicitly bypass the problem that the `_PROTECTED_KEYS` check solves. The problem only happens in the default copying implementation, so the problematic code in the CPython `copy` implementation is never executed if you override other methods that change the copying process.

Long version: The default behavior when you do `copy.copy(obj)` is basically something like `new_obj = type(obj).__new__(); for k in obj.__dict__: getattr(new_obj, k)`. However, Python objects offer multiple different hooks that the user can implement to customize how the object is copied. Therefore, when `copy.copy(obj)` is invoked, before the `for` loop I wrote above is run Python has to introspect `obj` for the presence of these other methods, which would indicate that copying should proceed by calling those rather than the default copying of `__dict__`. The recursion problem arises because looking up whether these methods are defined eventually leads to calling `__getattr__` on the object after it is created with `__new__`; since `__init__` is never called, the special `_data` attribute (for example) that would normally exist on the object when created via normal means is not defined, so the object is in an invalid state where `__getattr__` recursively searches for `self._data`. 

You can avoid this problem if you actually define one (or more) of those special methods. Python has a pretty complex and poorly documented hierarchy of different functions that can be used to modify the way that both copying and pickling of an object works; it turns out that many (but not all) functions are shared between these protocols. The only place I know to find the full dispatch sequence is by actually inspecting the CPython source code. Here's a rough summary of the important points, though. To customize the behavior of `copy`, the first thing that is used is the `__copy__` method. If it exists, then that's all that happens. If that is not defined, then `__getstate__` is invoked (if it exists) on the object being copied to get some data representing the "state" of the object. Then, once the new object is created (via `__new__`, with `__init__` never called), the new object's `__setstate__` method is called with the state generated by `__getstate__`. If neither `__getstate__` nor `__setstate__` is defined, the default copying of `__dict__` happens.

All of the methods I mentioned above (and a few others like `__getnewargs_ex__`) are part of Python's copying protocol, which is fundamentally implemented by another dunder method `__reduce__` that calls `__copy__` and `__getstate__` in the order I described above. You can instead override `__reduce__` itself to completely customize the process, but its signature is pretty cumbersome. That method must return a tuple that contains, at minimum, the class constructor to call (e.g. `cudf.DataFrame`), along with the arguments necessary to reconstruct this object. However, that tuple can also contain other items that allow you to customize things like setting the state of the object after initialization if there is state that cannot be controlled via the constructor alone. It's a pretty heavy hammer, though, and I'd avoid it unless absolutely necessary.

In this case, I would view any of these solutions as somewhat fragile because we're using this mixin as part of an inheritance tree with classes (like `Serializable`) that also may modify these methods, so it seems inadvisable to rely on subclasses not to override this behavior in problematic ways. As a result, the `_PROTECTED_KEYS` approach seems like the safer choice.

Hopefully that answers everything you could ever want to know about Python copying :)

Reply:
This explanation of python copy mechanism is super thorough. Thank you for this. I have to admit the first time I read the docstring I thought it discusses two different issues. But now it makes more sense. 

So to put it in perspective, if user calls `copy.copy` for one of cuDF object, it may cause a recursion error. Because in the new object, `__init__` was not invoked and `_PROTECTED_KEYS` is not yet defined, so `__getattr__` is called again.

However, cuDF/Pandas users are more advised to use the defined `copy` method on these objects, such as `Series.copy`, `DataFrame.copy` to perform copy, since these objects has underlying buffer that requires custom memory handling. It would be nice to see how we can adapt them into standard Python copy protocol - but an obvious counter argument is, if Python copy protocol is not well documented, perhaps we should discourage user from using `copy.copy` explicitly on cuDF objects? If the protocol is not well documented, it's not part of the language spec but implementation detail and should not be dependent on.

Besides top level objects that requires `GetAttrGetItemMixin`, is it possible there are internal objects that requires this as well? How should we consider python copy protocol to them? My personal take is we should discourage use of `copy.copy` unless deemed perfectly safe to do so, (against trivial objects, for example).

Reply:
Re docstring: personally I'm happy to read detailed design level ideas laid out in docstrings. Very often the issue is not detailed enough :). But to me what's in the docstring now seems
1. Either too succinct and contains some unstated assumptions (`that might also override serialization`, this happens when subclasses also overrides the serialization protocol, but not accounting for all parent classes when doing so. But why?). 
2. Or too detailed to account for the fact of `recursion error prevention`.

Reply:
I've reread the docstring and I definitely agree that I could do a better job of separating implementation details from what inheriting classes need to know in order to use it properly, which I think is the most important dividing line for developer documentation. I've tried to address this in the most recent rewrite, where I focus the docstring on telling developers what they need to do to safely use this class, and added a longer inline comment inside the class explaining the details of why this fix is necessary. Let me know if you think this is better.

Reply:
On the topic of whether or not `copy.copy` is safe, the expected behavior of `copy` and `pickle` methods is well-defined by the standard. Furthermore, the different ways to override these processes I mentioned above are more than just an implementation detail, they are actually promised based on [the various PEPs mentioned here](https://docs.python.org/3/library/pickle.html#data-stream-format). The problem is that none of these behaviors have good documentation. If you write a class that needs to customize the behavior of copying, deep copying, or pickling, the only good sources of information I know of for determining the best way to do this are the various PEPs linked in the pickle documentation linked above, or the CPython source code itself. In other words, this doesn't have any impact on users, but it can make developers lives very hard :) The fact that redefining other dunder methods like `__getattr__` can have unexpected effects on these protocols is I think just a fact of life that we have to live with.

Reply:
Re docstring: it looks very nice to me now. Thanks more making the rewrite!